theano/gof/cmodule.py

```diff
@@ -2310,5 +2319,314 @@ def print_command_line_error():
+        numpy_ver = [int(n) for n in numpy.__version__.split('.')[:2]]
+
+        # numpy 1.7 deprecated the following macro but the new one didn't
+        # existed in the past
+        if bool(numpy_ver < [1, 7]):
+            cxxflags.append("-DNPY_ARRAY_ENSUREARRAY=NPY_ENSUREARRAY")
+            cxxflags.append("-DNPY_ARRAY_ENSURECOPY=NPY_ENSURECOPY")
+            cxxflags.append("-DNPY_ARRAY_ALIGNED=NPY_ALIGNED")
+            cxxflags.append("-DNPY_ARRAY_WRITEABLE=NPY_WRITEABLE")
+            cxxflags.append("-DNPY_ARRAY_UPDATE_ALL=NPY_UPDATE_ALL")
+            cxxflags.append("-DNPY_ARRAY_C_CONTIGUOUS=NPY_C_CONTIGUOUS")
+            cxxflags.append("-DNPY_ARRAY_F_CONTIGUOUS=NPY_F_CONTIGUOUS")
+
+        # Platform-specific flags.
+        # We put them here, rather than in compile_str(), so they en up
+        # in the key of the compiled module, avoiding potential conflicts.
+
+    
+        # Figure out whether the current Python executable is 32
+        # or 64 bit and compile accordingly. This step is ignored for
+        # ARM (32-bit and 64-bit) architectures in order to make
+        # Theano compatible with the Raspberry Pi, Raspberry Pi 2, or
+        # other systems with ARM processors.
+        """
+        if (not any(['arm' in flag for flag in cxxflags]) and
+                not any(arch in platform.machine() for arch in ['arm', 'aarch'])):
+            n_bits = local_bitwidth()
+            cxxflags.append('-m%d' % n_bits)
+            _logger.debug("Compiling for %s bit architecture", n_bits)
+        """
+        if sys.platform != 'win32':
+            # Under Windows it looks like fPIC is useless. Compiler warning:
+            # '-fPIC ignored for target (all code is position independent)'
+            cxxflags.append('-fPIC')
+
+        if sys.platform == 'win32' and local_bitwidth() == 64:
+            # Under 64-bit Windows installation, sys.platform is 'win32'.
+            # We need to define MS_WIN64 for the preprocessor to be able to
+            # link with libpython.
+            cxxflags.append('-DMS_WIN64')
+
+        if sys.platform == 'darwin':
+            # Use the already-loaded python symbols.
+            cxxflags.extend(['-undefined', 'dynamic_lookup'])
+
+        return cxxflags
+
+    @staticmethod
+    def try_compile_tmp(src_code, tmp_prefix='', flags=(),
+                        try_run=False, output=False):
+        return Compiler._try_compile_tmp(src_code, tmp_prefix, flags,
+                                         try_run, output,
+                                         theano.config.cxx)
+
+    @staticmethod
+    def try_flags(flag_list, preambule="", body="",
+                  try_run=False, output=False):
+        return Compiler._try_flags(flag_list, preambule, body, try_run, output,
+                                   theano.config.cxx)
+
+    @staticmethod
+    def compile_str(module_name, src_code, location=None,
```

Comment:
I agree with that, much of the code for ICC_Compiler seems like a copy and paste from GCC_Compiler.  This is bad because it will be hard to maintain in the future and ICC_Compiler is much more likely to fall behind since there are less people using it.

There should be a refactor of the compiler interface so that multiple implementation can share as much code as possible (things like default definitions, compilation i/o, ...).